timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java

```diff
@@ -44,92 +38,43 @@ long lastVersion() {
         return nextSequence - 1;
     }
 
-    /**
-     * Adds an event to the sliding window. Assigns a unique sequence to the event.
-     *
-     * Note on concurrency:
-     * 1. Each write to buffer is followed by a write to nextSequence, which is volatile.
-     */
-    synchronized void add(LockWatchEvent.Builder eventBuilder) {
+    void add(LockWatchEvent.Builder eventBuilder) {
         LockWatchEvent event = eventBuilder.build(nextSequence);
         buffer[LongMath.mod(nextSequence, maxSize)] = event;
         nextSequence++;
     }
 
-    synchronized void finalizeAndAddSnapshot(long startVersion, LockWatchCreatedEventReplayer eventReplayer) {
-        Optional<List<LockWatchEvent>> remaining = getFromVersion(startVersion);
-        if (remaining.isPresent()) {
-            remaining.get().forEach(eventReplayer::replay);
-            add(LockWatchCreatedEvent.builder(eventReplayer.getReferences(), eventReplayer.getLockedDescriptors()));
-        }
+    boolean hasNextEvents(long version) {
+        return !validateVersion(version);
```

Comment:
sanity check: Is hasNextEvents() queried _at_ the last version intentionally true?

non-actionable: this is basically `hasEventsFromHereToLastKnown` though that is kind of verbose so just `hasNextEvents` is fine.

Reply:
Yes, if you present the last version, that means you're up to date, and we're return you last version and empty list of events. This method is used to calculate if the query to get the client up to date can be satisfied from the log, or else we need to calculate a snapshot.